apis/Google.Devtools.AspNet/Google.Devtools.AspNet/CloudTrace/RateLimiter.cs

```diff
@@ -0,0 +1,98 @@
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Diagnostics;
+
+namespace Google.Devtools.AspNet
+{
+    /// <summary>
+    /// A simple rate limiter that will allow tracing based on a given QPS.
+    /// </summary>
+    internal sealed class RateLimiter
+    {
+        // A mutex to protect the rate limiter instance.
+        private static object _instanceMutex = new object();
+
+        // The single rate limiter instance.
+        private static RateLimiter _instance;
+
+        // A mutex to handle access.
+        private object _mutex = new object();
+
+        // The amount of time that must be waited before allowing tracing.
+        private readonly long _fixedDelayMillis;
+
+        // A stopwatch to manage time between events.
+        private readonly Stopwatch _stopWatch;
+
+        // The last time tracing was allowed.
+        private long _lastCallMillis;
+
+        /// <summary>
+        /// Gets the instance of the <see cref="RateLimiter"/>.  The first request will set the
+        /// QPS of the rate limiter.  All subsequent request's QPS will not change the QPS of the rate limiter.
+        /// </summary>
+        public static RateLimiter GetInstance(double qps)
+        {
+            if (_instance == null)
+            {
+                lock (_instanceMutex)
+                {
+                    if (_instance == null)
+                    {
+                        _instance = new RateLimiter(qps);
+                    }
+                }
+            }
+            return _instance;
+        }
+
+        private RateLimiter(double qps) {
+            GaxPreconditions.CheckArgument(qps > 0, nameof(qps), "qps must be greater than 0");
+
+            _stopWatch = new Stopwatch();
+            _stopWatch.Start();
+
+            _lastCallMillis = 0;
+            _fixedDelayMillis = Convert.ToInt64(TimeSpan.FromSeconds(1 / qps).TotalMilliseconds);
+        }
+
+        /// <summary>
+        /// See if tracing is allowed.
+        /// </summary>
+        /// <returns>True tracing is allowed.</returns>
+        public bool CanTrace() {
+            if (CanTrace(_stopWatch.ElapsedMilliseconds))
+            {
+                lock (_mutex)
+                {
+                    long nowMillis = _stopWatch.ElapsedMilliseconds;
+                    if (CanTrace(nowMillis))
+                    {
+                        _lastCallMillis = nowMillis;
+                        return true;
+                    }
+                }
+            }
+            return false;
```

Comment:
I thought about this one some more this morning and you actually don't have to (and probably shouldn't) use the while loop for this kind of thing. If between getting _lastCallMillis and trying to set it, it has already been set by another thread, then that thread allowed a trace and this one shouldn't. The operation shouldn't be blocked while the thread checks again because it would only return true if this blocked long enough to hit the next interval. So you can probable use this instead:

```cs
var nowMillis = _stopWatch.ElapsedMilliseconds;
var lastCallMillis = _lastCallMillis;
return
    nowMillis - lastCallMillis > _fixedDelayMillis &&
    Interlocked.CompareExchange(ref _lastCallMillis, nowMillis, lastCallMillis) == lastCallMillis;
```